botocore/tokens.py

```diff
@@ -324,3 +327,25 @@ def load_token(self):
         return DeferredRefreshableToken(
             self.METHOD, self._refresher, time_fetcher=self._now
         )
+
+    def _resolve_source_profile(self, profiles, initial_profile_config):
+        profile_name = initial_profile_config.get('source_profile', None)
```

Comment:
When recursing through source profiles in addition to checking that the profile has a `source_profile`, it probably makes sense to check that a `role_arn` is set in the profile as well. The reason being that it makes it more strict when it comes to deciding whether to recurse through profiles to find an SSO configuration and limit this behavior only from the context of assume role configurations.

Specifically with the way we have it now, it is possible to chain profiles together to resolve an sso session solely on `source_profile` when retrieving and using the SSO token directly for authentication, which is not an intended interface:
```
[profile test]
source_profile = AdministratorAccess-XXXXXXXXXXXX
region = us-west-2

[profile AdministratorAccess-XXXXXXXXXXXX]
sso_session = dev
region = us-west-2

[sso-session dev]
sso_start_url = https://d-XXXXXXXXX.awsapps.com/start
sso_region = us-west-2
sso_registration_scopes = sso:account:access
```
Specifically, this configuration will allow some commands to be able to specify `--profile test` assuming the command uses bearer authentication, which is not a usage pattern we intend to support (instead the `test` profile would have to specify a `sso_session`)

Reply:
Thinking about this more, it may even make sense to make the behavior of recursing through profiles to locate the sso session name something that is opt in at the initialization of the `SSOTokenProvider` class. We can then toggle it on when instantiating the `SSOTokenProvider` as part of the credential provider chain, which would negate any potential side effects in the default auth token provider chain. Thoughts? I think the main point is that we should try to isolate this recursing logic to when we are using it as part of the assume role provider.